lib/model.js

```diff
@@ -347,13 +348,14 @@ Model.prototype.$__handleSave = function(options, callback) {
       }
       this.constructor.exists(where, optionsWithCustomValues).
         then((documentExists) => {
-          if (!documentExists) {
+          if (documentExists) {
+            const matchedCount = 1;
+            callback(null, { $where: where, matchedCount });
+          } else {
+            if (optimisticConcurrency) { this.$__.version = VERSION_INC; }
```

Comment:
This line is really what concerns me. I don't like the idea of bumping the version _after_ the operation is already done.

Reply:
I agree. 

The reason why I didn't set `this.$__.version = VERSION_ALL;`  is that the meaning of `VERSION_ALL` is not very clear to me as a newcomer(still not very clear), so I thought `VERSION_INC` is more appropriate. 

Now I know `VERSION_INC` means "to increase the version" instead of "version has been increased".

Let me close this PR and resend a new one based on the latest code since there's new code has been merged.